### DISCUSSION TITLE
NATS connector

### DISCUSSION THREAD

**Discussion by fernandocalsa** [Budibase/budibase#6266] [Ideas]
Hi everyone!

I've been working in the last few weeks with Budibase, it is amazing, and I'd like to contribute to improve it.

Right now I need to connect my Budibase with a NATS server. I think it would be great to add NATS as a new external connector, just like http.

In this case, NATS is different, as it is a pub-sub system:
1. Publishing messages: Just like a HTTP request where you won't have an answer. It can be triggered from a button or as part of any automation flow
2. Subscribe to topics: This is a bit different than any other data connector, and I can't see how it can match with other queries in other data connectors. It could be a trigger for an automation.

I think point 1 is easy to implement, and I'll try to send a PR if you think is valuable. But I think point 2 needs more discussion.

What do you think? Thank you!

**Comment by melohagan** [comment]
Hi @fernandocalsa 

Thank you very much for the kind words, and we appreciate your willingness to contribute! 😄 

Addressing your points:

1. If publishing can be done through a REST request, then it should be possible to use the External Data Connector as it is today without the need for code change. Have you tried it?
2. In a broader sense, subscription is not currently a concept in Budibase. I think the dev team would need to look at the general approach to handling subscribe events first, before opening this up to external contribution.

**Reply by fernandocalsa** [reply]
Hi @melohagan, thanks for your quick response!
1. NATS does not provide a native way to send a HTTP request and translate it into a NATS message, so I think I can not use the REST connector.
2. Yeah, I think the subscription phase is something that should be discussed somewhere else, and not only with NATS in mind, also with Websocket and other protocols.

Do you think a new connector with only the point 1 would be helpful? In my current project that could give us some options to extend our Budibase apps

**Reply by melohagan** [reply]
Ah I see. In that case adding a new connector would be wonderful! 

To be consistent with the other currently used connectors, this seems to be appropriate: https://github.com/nats-io/nats.js/#services-requestreply

I also found this guide to be helpful: https://docs.budibase.com/docs/writing-your-own-external-data-connector
